staging/src/k8s.io/apiserver/pkg/admission/options/authentication.go

```diff
@@ -0,0 +1,38 @@
+/*
```

Comment:
I am against this move of all the webhook specific constructs here. We leak complexity that should only be inside the webhook implementation. Instead, we should do the opposite: move even the initializer interfaces **into** the webhook package and find a way to abstract the webhook-specific initialization.

@deads2k ^^

Reply:
> I am against this move of all the webhook specific constructs here. We leak complexity that should only be inside the webhook implementation. Instead, we should do the opposite: move even the initializer interfaces into the webhook package and find a way to abstract the webhook-specific initialization.

I agree.  Only one API server out of every one created is going to be able to use a service resolver.  And a similar thing to the the clientconfig, it's a very rare use-case.  One that we can make possible, but not one that should leak out.

Reply:
To be a bit more concrete: can we have pkg/admission/plugin/webhook/initializers similar to the generic ones: https://github.com/kubernetes/kubernetes/blob/9d1e6d3e2cc25db8e07db446d00390059c8264f8/staging/src/k8s.io/apiserver/pkg/server/options/admission.go#L102 ? Or even just one custom initializer for everything that webhook needs?

Reply:
> We leak complexity that should only be inside the webhook implementation.

I don't get why the AuthenticationInfoResolver is only for the admission webhook. I thought every webhook (like the authz/n webhooks) should use it.

Reply:
> I don't get why the AuthenticationInfoResolver is only for the admission webhook. I thought every webhook (like the authz/n webhooks) should use it.

Other webhooks only connect to a single server.  Admission webhooks have to be able to connect to multiple servers.  The API level interface here is different.  Authentication and authorization aren't going to collapse onto this.

Reply:
I disagree about service resolver-- it's trivial for k8s.io/apiserver to offer a do-nothing one and I think it should. Makes it way easier for other apiservers to customize.

I somewhat agree about the auth info resolver, since it is so specific why is it even an interface at all? I don't think it needs to escape the webhook package boundaries?

Reply:
We can easily move them to pkg/admission/plugin/webhook/initializer. It does not restrict another apiserver to customize it by passing its own implementation. 

The whole mechanism of initializers was done in a way that plugins can easily provide their own initializers which are passed through the apiserver bootstrapping code. We don't loose anything here if we use that pattern. We keep the generic code simple and clean if we do.

As there are other PRs depending on this, I agree to do such a move separately in a follow-up. Let me create a PR on-top of this code here tomorrow.

Reply:
> I disagree about service resolver-- it's trivial for k8s.io/apiserver to offer a do-nothing one and I think it should. Makes it way easier for other apiservers to customize.
> 
> I somewhat agree about the auth info resolver, since it is so specific why is it even an interface at all? I don't think it needs to escape the webhook package boundaries?

We created admission plugin initializers to make it possible for someone composing with admission to inject configurable values without a need to leak those details outside the plugin and the particular composer.  While we do offer some plugin initializer methods in the genericapiserver, they are focused on the core use-cases.

A ServiceResolver can only be reliably used by the `kube-apiserver`.  It is extremely specific to a single composer (kube-apiserver) and shouldn't leak into the genericapiserver.  A similar argument exists for the authInfoResolver.  It is extremely specific to a single admission plugin and its likely that only kube-apiserver and openshift would ever even consider overriding it.

Reply:
Could `ServiceResolver` be used by other admission plugins? If so, the interface should be outside of the webhook.

Reply:
Specific to a single apiserver (which I don't agree with, but for the sake of conversation, will grant) is very different from specific to a single plugin.

Reply:
The generic apiserver is not the proper location to put plugin initializers that have single purpose audiences.  Given that there is no loss of power for keeping the details hidden, we should avoid exposure and if we get an indication that this is going to be common, there is no cost to later adding such an initializer.